File: FCFS.cpp
#include<bits/stdc++.h>
using namespace std;

struct Process
{
    int pID;    //? Process ID
    int aT;     //? Arrival Time
    int bT;     //? Burst Time
    int cT;     //? Completion Time
    int wT;     //? Waiting Time
    int tT;     //? Turnaround Time
};

bool compareProcess(Process p1, Process p2)
{
    return p1.aT < p2.aT;
}

int main()
{
    freopen("input.txt", "r", stdin);
    int _numProcesses;
    cout<<"Enter the number of processes: ";
    cin>> _numProcesses;

    vector<Process> _processes(_numProcesses);

    for(int i=0; i< _numProcesses; i++)
    {
        cout<<"Enter arrival time and burst time for process "<<i+1<<": ";
        cin>>_processes[i].aT >>_processes[i].bT;

        _processes[i].pID = i+1; //*Assinging the process ID's
    }
    fclose(stdin);

    //* Sort process based on arrival time
    sort(_processes.begin(), _processes.end(), compareProcess);

    //* Calculating CT, TT and WT
    int cT = 0;
    for(int i=0; i<_numProcesses; i++)
    {
        if(_processes[i].aT > cT)
        {
            cT = _processes[i].aT;
        }
        _processes[i].cT = cT + _processes[i].bT;
        _processes[i].tT = _processes[i].cT - _processes[i].aT;     //? Turnaround Time = Completiion Time - Arrival Time
        _processes[i].wT = _processes[i].tT - _processes[i].bT;     //? Waiting Time = Turnaround Time - Brust Time
        cT = _processes[i].cT;                                      //? Updating the completion time
    }


    //* Printing the process scheduling table
    cout<<"\n Process Scheduling Table: \n";
    cout<<"Process ID\tArrival Time\tBurst Time\tCompletion Time\tTurnaround Time\tWaiting Time\n";
    for(const Process& p : _processes)
    {
        cout<< p.pID << "\t\t"
            <<p.aT<<"\t\t"
            <<p.bT<<"\t\t"
            <<p.cT<<"\t\t"
            <<p.tT<<"\t\t"
            <<p.wT<<endl;
    }

    //* Calculating the average waiting time
    int totat_wT = 0;
    for(const Process& p: _processes)
    {
        totat_wT += p.wT;
    }
    float average_wT = static_cast<float>(totat_wT) / _numProcesses;    //? Average Waiting Time = Total Waiting Time / Number of Processes

    cout<<"Average Waiting Time: "<<average_wT<<endl;

    return 0;
}

File: fcfs.py
def fcfs_scheduling(processes):
    n = len(processes)
    
    # Initializing variables
    completion_time = [0] * n
    turnaround_time = [0] * n
    waiting_time = [0] * n
    start_time = [0] * n
    idle_time = 0
    
    # Sort processes by their arrival time
    processes.sort(key=lambda x: x[1])
    
    # Calculate completion time for each process
    for i in range(n):
        if i == 0 or completion_time[i-1] < processes[i][1]:
            completion_time[i] = processes[i][1] + processes[i][2]
            start_time[i] = processes[i][1]
        else:
            completion_time[i] = completion_time[i-1] + processes[i][2]
            start_time[i] = completion_time[i-1]
        
        # Calculate turnaround time: TAT = CT - AT
        turnaround_time[i] = completion_time[i] - processes[i][1]
        
        # Calculate waiting time: WT = TAT - BT
        waiting_time[i] = turnaround_time[i] - processes[i][2]
        
        # Calculate idle time
        if i > 0:
            idle_time += max(0, processes[i][1] - completion_time[i-1])
    
    # Print the results
    print("Process\tArrival Time\tBurst Time\tCompletion Time\tTurnaround Time\tWaiting Time")
    for i in range(n):
        print(f"{processes[i][0]}\t\t{processes[i][1]}\t\t{processes[i][2]}\t\t{completion_time[i]}\t\t{turnaround_time[i]}\t\t{waiting_time[i]}")
    print(f"\nTotal Idle Time: {idle_time}")

if __name__ == "__main__":
    n = int(input("Enter the number of processes: "))
    processes = []  # Each process is a tuple of (Process ID, Arrival Time, Burst Time)
    
    for i in range(n):
        process_id = i + 1
        arrival_time, burst_time = map(int, input(f"Enter Arrival Time and Burst Time for Process {process_id}: ").split())
        processes.append((process_id, arrival_time, burst_time))
    
    fcfs_scheduling(processes)
